File: ProgrammingChallenges#Book/Chapter 1/Minesweeper/solutions/Minesweeper.c
#include <stdio.h>

struct MineField {
    int n;
    int m;
    char field[100][100];
};

struct MineField minefield;

void printMineField();
int calcMines(int i, int j);
void resolveMinesweeper();

/* Main program */
int main () {
    int i, j, n, m, resultIndex = 1;
    char inputLine[100];

    while (scanf("%d %d", &minefield.n, &minefield.m) != EOF) {
        if (minefield.n == 0 && minefield.m == 0) {
            break;
        }

        if (resultIndex > 1) {
            printf("\n");
        }

        for (i = 0; i < minefield.n; i++) {
            scanf("%s", inputLine);
            for (j = 0; j < minefield.m; j++) {
                minefield.field[i][j] = inputLine[j];
            }
        }

        /* Resolve minesweeper */
        resolveMinesweeper();

        printf("Field #%d:\n", resultIndex);
        printMineField();
        resultIndex++;
    }

    return 0;
}


void printMineField() {
    int i, j;
    for (i = 0; i < minefield.n; i++) {
        for (j = 0; j < minefield.m; j++) {
            printf("%c", minefield.field[i][j]);
        }
        if (i != minefield.n -1) {
            printf("\n");
        }
    }
    printf("\n");
}

int calcMines(int i, int j) {
    int k, l, mines = 0;
    for (k = i-1; k <= i+1; k++) {
        for (l = j-1; l <= j+1; l++) {
            if (k >= 0 && k < minefield.n && l >= 0  && l < minefield.m) {
                if (minefield.field[k][l] == '*') {
                    mines++;
                }
            }
        }
    }

    return mines;
}

void resolveMinesweeper() {
    int i, j;
    for (i = 0; i < minefield.n; i++) {
        for (j = 0; j < minefield.m; j++) {
            if (minefield.field[i][j] == '.') {
                minefield.field[i][j] = calcMines(i, j) + 48;
            }
        }
    }
}

File: ProgrammingChallenges#Book/Chapter 1/Minesweeper/solutions/Minesweeper.py
from sys import stdin

class MineField:
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.field = [['' for x in range(m)] for y in range(n)]

        # Fill field
        for i in range(n):
            line = stdin.readline()
            for j in range(m):
                self.field[i][j] = line[j]

    def calcMines(self, i, j):
        mines = 0
        for k in range(i-1, i+2):
            for l in range(j-1, j+2):
                if (k >= 0 and k < self.n and l >= 0  and l < self.m):
                    if (self.field[k][l] == '*'):
                        mines += 1

        return mines;

    def resolveMinesweeper(self):
        for i in range(self.n):
            for j in range(self.m):
                if (self.field[i][j] == '.'):
                    self.field[i][j] = chr(self.calcMines(i, j) + 48)

    def __str__(self):
        out = "";
        for i in range(self.n):
            for j in range(self.m):
                out += self.field[i][j];

            if (i != self.n - 1):
                out+="\n";

        return out;


# Main program
def main():
    resultIndex = 1

    for line in stdin:
        n, m = map(int, line.split())

        if (n == 0 and m == 0):
            break;

        if (resultIndex > 1):
            print();

        minefield = MineField(n, m)
        # Resolve minesweeper
        minefield.resolveMinesweeper();

        print("Field #%d:" % (resultIndex))
        print(minefield)
        resultIndex += 1

main()